config/configtls/configtls.go

```diff
@@ -201,22 +202,97 @@ func (c TLSSetting) loadTLSConfig() (*tls.Config, error) {
 	}, nil
 }
 
-func (c TLSSetting) loadCert(caPath string) (*x509.CertPool, error) {
-	caPEM, err := os.ReadFile(filepath.Clean(caPath))
+func (c TLSSetting) loadCertPool() (*x509.CertPool, error) {
+	// There is no need to load the System Certs for RootCAs because
+	// if the value is nil, it will default to checking against th System Certs.
+	var err error
+	var certPool *x509.CertPool
+
+	switch {
+	case len(c.CAFile) != 0 && len(c.CAPem) != 0:
+		return nil, fmt.Errorf("failed to load CA CertPool: CA File and PEM cannot both be provided")
+	case len(c.CAFile) != 0:
+		// Set up user specified truststore from file
+		certPool, err = c.loadCertFile(c.CAFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load CA CertPool File: %w", err)
+		}
+	case len(c.CAPem) != 0:
+		// Set up user specified truststore from PEM
+		certPool, err = c.loadCertPem(c.CAPem)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load CA CertPool PEM: %w", err)
+		}
+	}
+
+	return certPool, nil
+}
+
+func (c TLSSetting) loadCertFile(caPath string) (*x509.CertPool, error) {
+	caPEM, err := ioutil.ReadFile(filepath.Clean(caPath))
 	if err != nil {
 		return nil, fmt.Errorf("failed to load CA %s: %w", caPath, err)
 	}
 
+	return c.loadCertPem(caPEM)
+}
+
+func (c TLSSetting) loadCertPem(caPEM []byte) (*x509.CertPool, error) {
 	certPool := x509.NewCertPool()
 	if !certPool.AppendCertsFromPEM(caPEM) {
-		return nil, fmt.Errorf("failed to parse CA %s", caPath)
+		return nil, fmt.Errorf("failed to parse CA %s", caPEM)
 	}
 	return certPool, nil
 }
 
+func (c TLSSetting) loadCertificate() (tls.Certificate, error) {
+	hasCertFile := len(c.CertFile) != 0
+	hasCertPem := len(c.CertPem) != 0
+	hasKeyFile := len(c.KeyFile) != 0
+	hasKeyPem := len(c.KeyPem) != 0
+	hasCert := hasCertFile || hasCertPem
+	hasKey := hasKeyFile || hasKeyPem
+
+	// Validate the incoming parameters.
+	switch {
```

Comment:
This can probably be simplified if you allow inline/file based mixing and matching.